numpy/core/src/multiarray/scalarapi.c

```diff
@@ -141,7 +141,18 @@ scalar_value(PyObject *scalar, PyArray_Descr *descr)
             return (void *)PyString_AS_STRING(scalar);
         }
         if (_CHK(Unicode)) {
-            return (void *)PyUnicode_AS_DATA(scalar);
+            /* Treat this the same as the NPY_UNICODE base class */
+
+            /* lazy initialization, to reduce the memory used by string scalars */
+            if (PyArrayScalar_VAL(scalar, Unicode) == NULL) {
+                Py_UCS4 *raw_data = PyUnicode_AsUCS4Copy(scalar);
+                if (raw_data == NULL) {
+                    return NULL;
+                }
+                PyArrayScalar_VAL(scalar, Unicode) = raw_data;
+                return (void *)raw_data;
+            }
+            return PyArrayScalar_VAL(scalar, Unicode);
```

Comment:
This is the same code as used above (line 74 or so) so the stanza itself is OK. The only other thing I could think of is to refactor `scalar_value()` to move the function itself onto the descriptor as a `__getscalar__` slot in the dtype refactoring.

Reply:
Yeah, we could do that and probably even should. I do not like `scalar_value()` as a requirement for scalars in general though.  And, I do not like flexible dtypes being special here, so would rather if we just return a `void *`.  In either case, Chuck probably wants to backport this, so we should put this in as is.

`scalar_value` (or something similar) is a neat idea.  But ultimately, it is mainly a slightly faster access to the value compared to `setitem` (i.e. one less copy).  The main problem for me is, that I do not really like it as a requirement on scalars (it technically is not even now), even if most scalars should be able to provide it easily.